sys/aarch64/board.c

```diff
@@ -8,66 +8,109 @@
 #include <sys/context.h>
 #include <sys/interrupt.h>
 #include <sys/kasan.h>
-#include <aarch64/atags.h>
+#include <sys/fdt.h>
+#include <sys/dtb.h>
 #include <aarch64/mcontext.h>
 #include <aarch64/vm_param.h>
+#include <aarch64/pmap.h>
 
-extern uint8_t _atags[];
-
-static int count_atags(void) {
-  atag_tag_t *atags = (atag_tag_t *)_atags;
-  int ntokens = 0;
-  atag_tag_t *atag;
-  ATAG_FOREACH(atag, atags) {
-    uint32_t tag = ATAG_TAG(atag);
-    if (tag == ATAG_MEM) {
-      ntokens++;
-    } else if (tag == ATAG_INITRD2) {
-      ntokens += 2;
-    } else if (tag == ATAG_CMDLINE) {
-      ntokens += cmdline_count_tokens(atag->tag_cmd.command);
-    }
-  }
-  return ntokens;
+/* Return offset of path at dtb or die. */
+static int dtb_offset(void *dtb, const char *path) {
+  int offset = fdt_path_offset(dtb, path);
+  if (offset < 0)
+    panic("Failed to find offset at dtb!");
+  return offset;
 }
 
-static void process_atags(char **tokens, kstack_t *stk) {
-  atag_tag_t *atags = (atag_tag_t *)_atags;
-  char buf[32];
+static uint32_t process_dtb_memsize(void *dtb) {
+  int offset = dtb_offset(dtb, "/memory@0");
+  int len;
+  const uint32_t *prop = fdt_getprop(dtb, offset, "reg", &len);
+  /* reg contains start (4 bytes) and end (4 bytes) */
+  if (prop == NULL || (size_t)len < 2 * sizeof(uint32_t))
+    panic("Failed to get memory size from dtb!");
 
-  atag_tag_t *atag;
-  ATAG_FOREACH(atag, atags) {
-    uint32_t tag = ATAG_TAG(atag);
-    if (tag == ATAG_MEM) {
-      snprintf(buf, sizeof(buf), "memsize=%d", atag->tag_mem.size);
-      tokens = cmdline_extract_tokens(stk, buf, tokens);
-    } else if (tag == ATAG_INITRD2) {
-      snprintf(buf, sizeof(buf), "rd_start=%d", atag->tag_initrd.start);
-      tokens = cmdline_extract_tokens(stk, buf, tokens);
-      snprintf(buf, sizeof(buf), "rd_size=%d", atag->tag_initrd.size);
-      tokens = cmdline_extract_tokens(stk, buf, tokens);
-    } else if (tag == ATAG_CMDLINE) {
-      tokens = cmdline_extract_tokens(stk, atag->tag_cmd.command, tokens);
-    }
-  }
+  return fdt32_to_cpu(prop[1]);
+}
+
+static int process_dtb_rd_start(void *dtb) {
+  int offset = dtb_offset(dtb, "/chosen");
+  int len;
+  const uint32_t *prop = fdt_getprop(dtb, offset, "linux,initrd-start", &len);
+  if (prop == NULL || (size_t)len < sizeof(uint32_t))
+    panic("Failed to get initrd-start from dtb!");
+
+  return fdt32_to_cpu(*prop);
+}
+
+static int process_dtb_rd_size(void *dtb) {
+  int offset = dtb_offset(dtb, "/chosen");
+  int len;
+  const uint32_t *prop = fdt_getprop(dtb, offset, "linux,initrd-end", &len);
+  if (prop == NULL || (size_t)len < sizeof(uint32_t))
+    panic("Failed to get initrd-start from dtb!");
+
+  return fdt32_to_cpu(*prop) - process_dtb_rd_start(dtb);
```

Comment:
Not sure if it is necessary, but could you add an assertion that `end >= start`?

Reply:
It is already asserted in physmem.